clippy_lints/src/manual_let_else.rs

```diff
@@ -167,6 +169,50 @@ fn emit_manual_let_else(
     );
 }
 
+/// Returns whether the given let pattern and else body can be turned into a question mark
+///
+/// For this example:
+/// ```ignore
+/// let FooBar { a, b } = if let Some(a) = ex { a } else { return None };
+/// ```
+/// We get as parameters:
+/// ```ignore
+/// pat: Some(a)
+/// else_body: return None
+/// ```
+
+/// And for this example:
+/// ```ignore
+/// let Some(FooBar { a, b }) = ex else { return None };
+/// ```
+/// We get as parameters:
+/// ```ignore
+/// pat: Some(FooBar { a, b })
+/// else_body: return None
+/// ```
+
+/// We output `Some(a)` in the first instance, and `Some(FooBar { a, b })` in the second, because
+/// the question mark operator is applicable here. Callers have to check whether we are in a
+/// constant or not.
+pub fn pat_and_expr_can_be_question_mark<'a, 'hir>(
```

Comment:
+1

Reply:
The `question_mark` lint is implemented right now as a visitor on expressions, while the `manual_let_else` lint looks for statements. It makes a lot of sense for both to do it this way, because `let v = if let Some(v) = ex { v } else { return None };` is a statement while for `question_mark` it looks for stuff like `if let Some(v) = ex { v } else { return None }`, which is an expression, and it doesn't matter if that expression is put into a local or is the start of a method chain, directly a function parameter, etc.

So while it's possible to merge the two lints into one pass, likely that would just similarly call the same function two times, just once through a check on statements, once through a check on expressions.

The only way I could think of unifying is that one changes the `manual_let_else` lint to look for expressions, and then make it obtain the parent node, and if the parent node is a `let` statement, it would continue from there.

Do you think this is preferable to having two passes?

Reply:
I would do this way simpler:

```rust
fn check_let_some_else_return_none(..) -> bool
```

should return a `bool` whether the lint was triggered.

in the `check_stmt` part of the unified pass, you would then write the code like this:

```rust
if !check_let_some_else_return_none(...) {
    check_manual_let_else(...);
}
```

And in `check_manual_let_else` you put the code that currently is in the `LateLintPass` impl of `check_stmt` of the `manual_let_else` lint.

Maybe it makes sense to include those lints in the `matches` module. But I'm not 100% convinced if this is the best location for it.

Reply:
The `check_let_some_else_return_none` function does only check for `let...else`, while the `check_manual_let_else` function checks for `let foo = if let () ...` patterns (and the match equivalent). Therefore, there is no overlap here. The overlap with the question mark lint would happen with the `<QuestionMark as LateLintPass>::check_expr` function, which takes an expression instead of a statement.

Reply:
I have pushed a commit to put the two things into one pass. I feel it's better without this merging, but maybe you like it more than I do.